### PR TITLE
Split cancellation into a two step process

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
Previously the status messages (`NODE_COMPLETED_JOB`, `NODE_FAILED_ARRAY_TASK`) all use to signify the job/task was completed and the node is likely de-allocated. This is inconsistent with the allocation method which has previously been converted into a two step process with `JOB_ALLOCATED` and `RUN_SCRIPT`/`RUN_STEP`.

The other problem is only a single node will report back most of the `NODE_*_JOB`/`NODE_*_ARRAY_TASK` statuses, but all the nodes within the allocation received a `JOB_ALLOCATED` request.

The fix is to require all `JOB_ALLOCATED` requests to be explicitly relinquished with another message. This explicit message can either be `JOB_CANCELLED` or `JOB_DEALLOCATED`. These messages are sent to every daemon in the allocation as any one of them maybe running a job step.

When a daemon receives `JOB_CANCELLED` or `JOB_DEALLOCATED`, it will shutdown the job's running processes with `SIGTERM` or a natural exit respectively. Once all the the running processes have stopped, the `daemon` will respond `NODE_DEALLOCATED`.

Finally the scheduler processes the `NODE_DEALLOCATED` messages to determine when the job has finished and which nodes are available to be reassigned.

### CODE REVIEW

File: app/websocket_app.rb [hunk 1]
```diff
@@ -72,20 +76,21 @@ def call(message)
     else
       Async.logger.info("Unknown message #{message}")
     end
-  rescue
+  rescue => e
     Async.logger.info("Error processing message #{$!.message}")
+    Async.logger.debug(e.full_message)
```
Comment (benarmston) [inline app/websocket_app.rb]: We're using both `e` and `$!` to refer to the exception in the same rescue block.  We should be consistent in our use especially inside of a single rescue block and ideally across the entire code base.  FWIW, I prefer `$!` as 1) it always refers to the exception, whereas `e` *could* refer to any object at all; and 2) we're already using `$!` across the code base.

File: lib/flight_scheduler/deallocation/job.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+          command: 'JOB_DEALLOCATED',
+          job_id: @job.id,
+        })
+        connection.flush
+        Async.logger.debug("Job deallocation for #{@job.id} sent to #{target_node.name}")
+      end
+
+    rescue => e
+      Async.logger.error("Error deallocating job #{@job.id}: #{$!.message}")
+      Async.logger.debug e.full_message
```
Comment (benarmston) [inline lib/flight_scheduler/deallocation/job.rb]: Same here about accessing the exception.  Let's be consistent between `$!` and `e`.

File: lib/flight_scheduler/event_processor.rb [hunk 3]
```diff
@@ -110,68 +75,93 @@ def allocate_resources_and_run_jobs
   def node_completed_job(node_name, job_id)
     Async.logger.info("Node #{node_name} completed job #{job_id}")
     allocation = FlightScheduler.app.allocations.for_job(job_id)
+
+    # Ignore responses which have already been deallocated
+    # NOTE: There is a race condition between the state and deallocation messages
```
Comment (benarmston) [inline lib/flight_scheduler/event_processor.rb:80]: Can you explain the race condition please?

File: lib/flight_scheduler/event_processor.rb [hunk 4]
```diff
@@ -190,4 +180,47 @@ def job_step_failed(node_name, job_id, step_id)
...
   module_function :job_step_failed
+
+  def cancel_job(job)
+    case job.job_type
+    when 'JOB'
+      cancel_batch_job(job)
+    when 'ARRAY_JOB'
+      cancel_array_job(job)
+    else
+      Async.logger.error("Can not cancel #{job.job_type} #{job.id}")
```
Comment (benarmston) [inline lib/flight_scheduler/event_processor.rb]: Typo: `Cannot` instead of `Can not`.

**Review by benarmston — CHANGES_REQUESTED** [review]
This looks good.  It looks like you've hit every issue with job cancellation.  I've added a comment about consistency in referring to exceptions, and I'd like you to explain the race condition that you mention.  Other than those, I think this is pretty much there.

File: lib/flight_scheduler/event_processor.rb (hunk 3 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/event_processor.rb:80]: I haven't fully tested them out, so the following might be wrong in parts (needs confirmation). However the following may occur:

The `daemon` uses two asynchronous task to send `NODE_COMPLETED_JOB` (insert other status message) and `NODE_DEALLOCATED`. The first is waiting for the process to die before removing the runner. The latter waits for the removal of all runners. However as everything is asynchronous its hard to predict the exact execution flow.

I was particularly having the issue when the `JOB_CANCELLED` method was partially implemented, however I believe this particular case has been resolved.

In general its probably best not to assume one message will be received before the other. Furthermore using the `AllocationRegistry` as a pseudo "Job Registry" is part of the above problem.

File: lib/flight_scheduler/event_processor.rb (hunk 4 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/event_processor.rb]: https://github.com/openflighthpc/flight-scheduler-controller/pull/22/commits/9027e157563ccbc453d6b8b4f33e52af6c224e4a

File: lib/flight_scheduler/deallocation/job.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/deallocation/job.rb]: https://github.com/openflighthpc/flight-scheduler-controller/pull/22/commits/9027e157563ccbc453d6b8b4f33e52af6c224e4a

File: app/websocket_app.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply app/websocket_app.rb]: https://github.com/openflighthpc/flight-scheduler-controller/pull/22/commits/9027e157563ccbc453d6b8b4f33e52af6c224e4a

File: lib/flight_scheduler/event_processor.rb (hunk 3 above)
Comment (benarmston) [reply lib/flight_scheduler/event_processor.rb:80]: OK. Thanks for the response.  I think it'll be worth looking at this again if/when we get a job registry/job persistence.

**Comment by benarmston** [comment]
Thanks for the changes. This LGTM.

**Pull request merged**